Logging

ageorgou opened the pull request:
This will introduce some basic logging functionality to help us with debugging, and will partly address #20. 

Things that we should log:
- [x] Users requesting actions (start, stop, ...)
- [x] In particular, information about deletion of machines, such as the choice of whether deletion is carried out through Terraform or the Azure API, or whether deployment had been completed before deletion was requested
- [x] What Terraform commands are run
- [x] Communication with Azure
- [x] Machine creation
- [x] User management (registration, adding keys)
- [ ] Errors thrown from Celery, Flask, TF ... ?

Things that are not logged because they are recorded elsewhere:
- Terraform output (stored in the deployment log in the database)
- Basic Celery information, i.e. each task being sent, received and completed (in the Celery log, which on local builds is on stdout)

The logging level is just a hunch at the moment. My intention was to log most actions as INFO, the lower-level Azure communication as DEBUG, and unexpected things as WARNING/ERROR.

We should also:
- [x] Decide on a format for the messages (e.g. do we want the code line?)
- [x] Fine tune the rolling file hander (e.g. size before rolling over), or switch to another file handler
- [x] Set up the logging from a config file, or at least have some flexibility depending on what configuration the Flask app is running in

ageorgou committed:
Set up a basic dedicated log

ageorgou committed:
Log attempts to manipulate hosts owned by others

It seems more correct to return a 403 error code rather than 404,
but maybe that is less secure? (it gives away information about
what host IDs exist under other uses)

ageorgou committed:
Log users' requests for operations on hosts

ageorgou committed:
Change old debug messages to log entries

ageorgou committed:
Log Terraform commands, communication with Azure

ageorgou committed:
Remove unused import

ageorgou committed:
Add messages for start and end of deployment

ageorgou committed:
Add messages about key management and login errors

ageorgou committed:
Log information about user management

raquelalegre commented on src/cloudlabs/app.py:
```diff
@@ -22,6 +24,17 @@ def create_app(config_name=None):
+        '%(asctime)s %(levelname)s: %(message)s '
+        '[in %(pathname)s:%(lineno)d]'
+    ))
+    app.logger.addHandler(handler)
```
Can this configuration be done on a separate config file? I've found that useful in the past and it might come in handy as the application grows so we can have separate loggers for different things (e.g. one for deployment related messages, one for user management, etc.). 

Maybe [this](https://stackoverflow.com/questions/11816236/how-to-config-flask-app-logger-from-a-configure-file?utm_medium=organic&utm_source=google_rich_qa&utm_campaign=google_rich_qa) and [this](https://gist.github.com/jakubczaplicki/c9077a482c2b195b7f3bcbee19dde441) help. 

[This other one](https://github.com/oracc/nammu/blob/development/resources/config/logging.yaml) is the one I did for Nammu for the `logging` python module, which might or might not be that relevant.

ageorgou replied:
Yes, I agree that we should. I wanted to get something started and then forgot about it. I'll do this, possibly after #72 is merged so we have the new config too.

ageorgou replied:
I just reread this and had misunderstood - I was thinking of doing it in `config.py` for each configuration of the Flask app (development, production, ...). Did you mean a separate file that controls the logging configuration as in Nammu?

ageorgou replied:
Now done similarly to Nammu.

raquelalegre commented on src/cloudlabs/azure_tools.py:
```diff
@@ -1,14 +1,34 @@
+        return helper
+    return wrap
+
+
```
👍 !

raquelalegre commented on src/cloudlabs/deployer/deployer.py:
```diff
@@ -83,8 +82,12 @@ def deploy(self, host):
         else:
             self._record_result(host, HostStatus.error, 'apply', process.returncode)
+            current_app.logger.error(
+                "Deployment of host %s failed (Terraform return code %s)",
```
Tiny typo - should be `returned`.

raquelalegre replied:
Ah no. It's the "return code". My bad :)

ageorgou committed:
Add more messages on deletion

ageorgou committed:
Use separate loggers for each part of the app

ageorgou committed:
Configure the new loggers

Note that we need to configure our own loggers (cloudlabs.*)
specifically, and not the root! This is because Celery overrides
the root logger configuration. It should be possible to turn that
off, but it seems it's not that simple. Setting the congiruation
for our family of loggers seems to work, however, and we only need
to set the "parent" logger (cloudlabs) rather than each logger
(e.g. cloudlabs.admin) explicitly.

ageorgou committed:
Configure logger from YAML file

This also makes some minor changes to the log format and settings.
NOTE: It's possible that the location of the log config file has
to be given in a different way, in case Flask is started from a
different location (eg through Apache). Will need to check this
on the servers!

ageorgou committed:
Make some style changes

ageorgou committed:
Remove accidental trailing ' from log entries

ageorgou commented:
Tests failing on 3.4 because of an unrelated update to the `secrestorage` package, which now requires Python>=3.5.

raquelalegre committed:
Travis tests run on Python 3.5 instead of 3.4 due to error introduced by secretstorage new version.

raquelalegre commented on src/cloudlabs/azure_tools.py:
```diff
@@ -9,6 +12,22 @@
+        return helper
+    return wrap
+
+
```
Awesome!

raquelalegre reviewed (commented):
Looks good! I think there's just propagating the errors form TF and Celery that is left doing, right? I want to try this branch also myself and then it's ready to merge.

ageorgou committed:
Forward terraform's stderr to log

ageorgou commented:
Celery information is stored in its own logs (on stdout/stderr if running locally, in `/var/log/celery/*.log` on the servers). It seems that Celery will try to set up its own configuration regardless of the user's setup, so I would prefer to avoid messing with that.

Terraform errors are now captured in both the log and the DB. I'm trying to see if we need to/can log anything from Flask.

ageorgou commented:
Note that this needs to be tweaked to run on the servers - in particular, either raising the logging level, or removing the logging call from `AzureTools.refresh()`, or the whole log will fill up from the auto-status-refresh task (although it may be better to see if we can avoid getting new credentials each time the refresh task runs).

ageorgou committed:
Disable refresh credentials log message

ageorgou commented:
Getting errors from Flask (e.g. exceptions) is doable but maybe not helpful:
- In development mode, the interactive display is nicer than a message in a log file. If the error is not produced from a view, it will be displayed on stderr anyway.
- In production, any Flask errors are (theoretically, at least) written in the Apache log.

Furthermore, the only way I found of doing it is not very clean (has to be done separately from the YAML configuration), and, if also affecting development mode, replaces the nice default display.

All in all, I'm not sure whether adding this is a good idea, unless I can find a better way of doing it.

ageorgou commented:
@raquel-ucl's suggestion: record celery task id's in the main log, for easy crossreferencing with the celery log.

This should be doable by examining the return value when we send a celery task.

ageorgou committed:
Record Celery task IDs when (re)starting, stopping

Also some small reorganisation, getting rid of some unreachable
code from a previous version. This referred to the "Not Implementd"
template which is now unused, but I have left it in the code in
case we want to introduce more features gradually in the future.
